Fix bug #7916, allowing `Reflexive` to have mode `! !` (#13969)
coq/coq: doc/sphinx/addendum/generalized-rewriting.rst

solution to the set of constraints generated by a rewrite and will not
try to find *all* the possible solutions to warn the user about them.

Compatibility with ssreflect's rewrite
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When calling ssreflect's :tacn:`rewrite` tactic with a lemma, a check is
> performed to see if the rewriting relation can be handled by generalized rewriting.
> This relies on the ``RewriteRelation A R`` typeclass, indexed by a type and relation,
> that registers relations handled by generalized rewriting.

Reviewer: ```suggestion
The :tacn:`rewrite (ssreflect)` tactic uses generalized rewriting when possible.
The ``RewriteRelation A R`` typeclass, indexed by a type and relation, identifies
the relations that generalized rewriting handles.
```

If that says what's needed here.

If some of this information applies only to the one tactic, maybe it should appear in the description of `rewrite` in the SSR chapter?

Author: We could backlink from the ssr chapter indeed.

The default instances of this class are the ``iff```, ``impl`` and ``flip impl``

> relations on ``Prop``, any declared ``Equivalence`` on a type ``A`` (including Leibniz equality),

Reviewer: ```suggestion
relations on ``Prop``, any declared ``Equivalence`` on a type ``A`` (including :term:`Leibniz equality`),
```

Hyperlink to the definition.

One can inform the rewrite tactic about morphisms and relations just
by using the typeclass mechanism to declare them using the :cmd:`Instance` and

> :cmd:`Context` commands. Any object of type ``Proper`` (the type of

Reviewer: I take it `Proper` is a `Type`?  Maybe at some point we can add a list of such predefined Types in the doc.
